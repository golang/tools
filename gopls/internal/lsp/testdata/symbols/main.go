package main

import (
	"io"
)

// Each symbol marker in this file defines the following information:
//  symbol(name, selectionSpan, kind, detail, id, parentID)
//    - name: DocumentSymbol.Name
//    - selectionSpan: DocumentSymbol.SelectionRange
//    - kind: DocumentSymbol.Kind
//    - detail: DocumentSymbol.Detail
//    - id: if non-empty, a unique identifier for this symbol
//    - parentID: if non-empty, the id of the parent of this symbol
//
// This data in aggregate defines a set of document symbols and their
// parent-child relationships, which is compared against the DocummentSymbols
// response from gopls for the current file.
//
// TODO(rfindley): the symbol annotations here are complicated and difficult to
// maintain. It would be simpler to just write out the full expected response
// in the golden file, perhaps as raw JSON.

var _ = 1

var x = 42 //@symbol("x", "x", "Variable", "", "", "")

var nested struct { //@symbol("nested", "nested", "Variable", "struct{...}", "nested", "")
	nestedField struct { //@symbol("nestedField", "nestedField", "Field", "struct{...}", "nestedField", "nested")
		f int //@symbol("f", "f", "Field", "int", "", "nestedField")
	}
}

const y = 43 //@symbol("y", "y", "Constant", "", "", "")

type Number int //@symbol("Number", "Number", "Class", "int", "", "")

type Alias = string //@symbol("Alias", "Alias", "Class", "string", "", "")

type NumberAlias = Number //@symbol("NumberAlias", "NumberAlias", "Class", "Number", "", "")

type (
	Boolean   bool   //@symbol("Boolean", "Boolean", "Class", "bool", "", "")
	BoolAlias = bool //@symbol("BoolAlias", "BoolAlias", "Class", "bool", "", "")
)

type Foo struct { //@symbol("Foo", "Foo", "Struct", "struct{...}", "Foo", "")
	Quux                    //@symbol("Quux", "Quux", "Field", "Quux", "", "Foo")
	W         io.Writer     //@symbol("W", "W", "Field", "io.Writer", "", "Foo")
	Bar       int           //@symbol("Bar", "Bar", "Field", "int", "", "Foo")
	baz       string        //@symbol("baz", "baz", "Field", "string", "", "Foo")
	funcField func(int) int //@symbol("funcField", "funcField", "Field", "func(int) int", "", "Foo")
}

type Quux struct { //@symbol("Quux", "Quux", "Struct", "struct{...}", "Quux", "")
	X, Y float64 //@symbol("X", "X", "Field", "float64", "", "Quux"), symbol("Y", "Y", "Field", "float64", "", "Quux")
}

type EmptyStruct struct{} //@symbol("EmptyStruct", "EmptyStruct", "Struct", "struct{}", "", "")

func (f Foo) Baz() string { //@symbol("(Foo).Baz", "Baz", "Method", "func() string", "", "")
	return f.baz
}

func _() {}

func (q *Quux) Do() {} //@symbol("(*Quux).Do", "Do", "Method", "func()", "", "")

func main() { //@symbol("main", "main", "Function", "func()", "", "")
}

type Stringer interface { //@symbol("Stringer", "Stringer", "Interface", "interface{...}", "Stringer", "")
	String() string //@symbol("String", "String", "Method", "func() string", "", "Stringer")
}

type ABer interface { //@symbol("ABer", "ABer", "Interface", "interface{...}", "ABer", "")
	B()        //@symbol("B", "B", "Method", "func()", "", "ABer")
	A() string //@symbol("A", "A", "Method", "func() string", "", "ABer")
}

type WithEmbeddeds interface { //@symbol("WithEmbeddeds", "WithEmbeddeds", "Interface", "interface{...}", "WithEmbeddeds", "")
	Do()      //@symbol("Do", "Do", "Method", "func()", "", "WithEmbeddeds")
	ABer      //@symbol("ABer", "ABer", "Field", "ABer", "", "WithEmbeddeds")
	io.Writer //@symbol("Writer", "Writer", "Field", "io.Writer", "", "WithEmbeddeds")
}

type EmptyInterface interface{} //@symbol("EmptyInterface", "EmptyInterface", "Interface", "interface{}", "", "")

func Dunk() int { return 0 } //@symbol("Dunk", "Dunk", "Function", "func() int", "", "")

func dunk() {} //@symbol("dunk", "dunk", "Function", "func()", "", "")
