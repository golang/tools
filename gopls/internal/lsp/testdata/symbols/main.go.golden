-- symbols --
x Variable 26:5-26:6
nested Variable 28:5-28:11
	nestedField Field 29:2-29:13
y Constant 34:7-34:8
Number Class 36:6-36:12
Alias Class 38:6-38:11
NumberAlias Class 40:6-40:17
Boolean Class 43:2-43:9
BoolAlias Class 44:2-44:11
Foo Struct 47:6-47:9
	Bar Field 50:2-50:5
	Quux Field 48:2-48:6
	W Field 49:2-49:3
	baz Field 51:2-51:5
	funcField Field 52:2-52:11
Quux Struct 55:6-55:10
	X Field 56:2-56:3
	Y Field 56:5-56:6
EmptyStruct Struct 59:6-59:17
(Foo).Baz Method 61:14-61:17
(*Quux).Do Method 67:16-67:18
main Function 69:6-69:10
Stringer Interface 72:6-72:14
	String Method 73:2-73:8
ABer Interface 76:6-76:10
	A Method 78:2-78:3
	B Method 77:2-77:3
WithEmbeddeds Interface 81:6-81:19
	ABer Field 83:2-83:6
	Do Method 82:2-82:4
	Writer Field 84:5-84:11
EmptyInterface Interface 87:6-87:20
Dunk Function 89:6-89:10
dunk Function 91:6-91:10

